src/main/kotlin/org/cascadebot/cascadebot/data/objects/CommandFilter.kt

```diff
@@ -0,0 +1,99 @@
+ */
+
+package org.cascadebot.cascadebot.data.objects
+
+import net.dv8tion.jda.api.EmbedBuilder
+import net.dv8tion.jda.api.entities.Member
+import net.dv8tion.jda.api.entities.Role
+import net.dv8tion.jda.api.entities.TextChannel
+import java.util.Collections
+
+class CommandFilter(val name: String) {
+
+    // Constructor for MongoDB
+    private constructor() : this("")
+
+    var type = FilterType.BLACKLIST
+    var operator = FilterOperator.AND
+
+    var enabled = true
+
+    val commands: MutableSet<String> = Collections.synchronizedSet(mutableSetOf())
+    val channelIds: MutableSet<Long> = Collections.synchronizedSet(mutableSetOf())
+    val userIds: MutableSet<Long> = Collections.synchronizedSet(mutableSetOf())
+    val roleIds: MutableSet<Long> = Collections.synchronizedSet(mutableSetOf())
+
+    fun evaluateFilter(command: String, channel: TextChannel, member: Member): FilterResult {
+        if (!enabled) {
+            return FilterResult.NEUTRAL
+        }
+        if (!commands.contains(command)) {
+            return FilterResult.NEUTRAL
+        }
+
+        var channelMatch = FilterMatch.NEUTRAL
+        // The channel condition is only considered if one or more channels have been added to the filter
+        if (channelIds.size != 0) {
+            channelMatch = if (channelIds.contains(channel.idLong)) FilterMatch.MATCH else FilterMatch.NOT_MATCH
+        }
+
+        var userMatch = FilterMatch.NEUTRAL
+        // The user condition is only considered if one or more users have been added to the filter
+        if (userIds.size != 0) {
+            userMatch = if (userIds.contains(member.idLong)) FilterMatch.MATCH else FilterMatch.NOT_MATCH
+        }
+
+        var roleMatch = FilterMatch.NEUTRAL
+        // The role condition is only considered if one or more roles have been added to the filter
+        if (roleIds.size != 0) {
+            roleMatch = if (member.roles.stream().map { obj: Role -> obj.idLong }.anyMatch { id: Long -> roleIds.contains(id) }) FilterMatch.MATCH else FilterMatch.NOT_MATCH
+        }
+
+        val combinedResult: Boolean = if (operator == FilterOperator.AND) {
+            // Check that all of the results are either MATCH or NEUTRAL
+            channelMatch != FilterMatch.NOT_MATCH && userMatch != FilterMatch.NOT_MATCH && roleMatch != FilterMatch.NOT_MATCH
+        } else {
+            // Check that any of the results are either MATCH or NEUTRAL
+            channelMatch != FilterMatch.NOT_MATCH || userMatch != FilterMatch.NOT_MATCH || roleMatch != FilterMatch.NOT_MATCH
+        }
+        return when (type) {
+            FilterType.WHITELIST -> if (combinedResult) FilterResult.ALLOW else FilterResult.DENY
+            FilterType.BLACKLIST -> if (combinedResult) FilterResult.DENY else FilterResult.ALLOW
+        }
+    }
+
+    val filterEmbed: EmbedBuilder
+        get() = EmbedBuilder()
+
+    /**
+     * Determines whether users who match this filter will be blocked or whitelisted
```

Comment:
this is just me being stupid, but maybe make a separate file for each of these enums? probably ott

Reply:
The answer there is.... not sure?